### ISSUE TITLE
rmw_fastrtps_* fails to publish anything after #738 [6507]

### ISSUE THREAD

**Issue by cottsay** [eProsima/Fast-RTPS#758]
Once #751 was resolved, our tests were able to run, which revealed that all attempts to publish data fail with this message:

```
Failed to publish log message to rosout: cannot publish data, at src/ros2/rmw_fastrtps/rmw_fastrtps_shared_cpp/src/rmw_publish.cpp:52, at src/ros2/rcl/rcl/src/rcl/publisher.c:257
```

Because so many tests were failing, I'm not able to give any build logs due to the jobs' build logs exceeding 20 GiB, but this is simple to reproduce locally by just running `ros2 run demo_nodes_cpp talker`.

Blocks ros2/ros2#793

**Comment by LuisGP** [comment]
Hi @cottsay 

We think we solved the issue. It was caused because the TopicAttributes were going out of scope. Due to #738  refactor we were storing (too much optimistically) a const ref to it instead of making a copy.

You may want to test it before it is merged.
